src/shared/sam/cli/samCliContext.ts

```diff
@@ -0,0 +1,82 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import { Logger } from '../../logger'
+import { SettingsConfiguration } from '../../settingsConfiguration'
+import { DefaultSamCliConfiguration } from './samCliConfiguration'
+import { DefaultSamCliProcessInvoker, SamCliProcessInvokerContext } from './samCliInvoker'
+import { SamCliProcessInvoker } from './samCliInvokerUtils'
+import { DefaultSamCliLocationProvider } from './samCliLocator'
+import { DefaultSamCliValidator, DefaultSamCliValidatorContext, SamCliValidator } from './samCliValidator'
+
+export interface SamCliContext {
+    validator: SamCliValidator,
+    invoker: SamCliProcessInvoker,
+}
+
+// Sam Cli Context is lazy loaded on first request to reduce the
+// amount of work done during extension activation.
+let samCliContext: SamCliContext | undefined
+let samCliContextInitialized: boolean = false
+
+// Components required to load Sam Cli Context
+let settingsConfiguration: SettingsConfiguration
+let logger: Logger
+
+export function initialize(
+    {
```

Comment:
Nit: Can just do `params: { ... }` instead of `{ ...params }: { ... }`. But since you're expanding params inside the method anyway, simpler to replace this:

```typescript
export function initialize(
    {
        ...params
    }: {
        settingsConfiguration: SettingsConfiguration,
        logger: Logger,
    }
) {
    settingsConfiguration = params.settingsConfiguration
    logger = params.logger

    samCliContext = undefined
    samCliContextInitialized = true
}
```

With this:

```typescript
export function initialize(
    { settingsConfiguration, logger }: {
        settingsConfiguration: SettingsConfiguration,
        logger: Logger,
    }
) {
    samCliContext = undefined
    samCliContextInitialized = true
}
```

Reply:
The current approach prevents shadowing the variables on line 27,28

Reply:
In that case, you can just do `params: { ... }` instead of `{ ...params }: { ... }`.